File: modules/gd-kasane/src/deformer_data.cpp
// SPDX-License-Identifier: MIT
#include "deformer_data.hpp"
#include "document_bridge.hpp"
#include <godot_cpp/core/class_db.hpp>
#include <godot_cpp/core/object.hpp>
#include <godot_cpp/classes/os.hpp>
using namespace godot;

namespace kasane_gd {
void KasaneDeformerData::_bind_methods() {
    ClassDB::bind_method(D_METHOD("is_valid"), &KasaneDeformerData::is_valid);
    ClassDB::bind_method(D_METHOD("get_id"), &KasaneDeformerData::get_id);
    ClassDB::bind_method(D_METHOD("snapshot"), &KasaneDeformerData::snapshot);
    ClassDB::bind_method(D_METHOD("get_angle_degrees"), &KasaneDeformerData::get_angle_degrees);
    ClassDB::bind_method(D_METHOD("set_angle_degrees", "value"), &KasaneDeformerData::set_angle_degrees);
    ClassDB::bind_method(D_METHOD("get_center"), &KasaneDeformerData::get_center);
    ClassDB::bind_method(D_METHOD("set_center", "value"), &KasaneDeformerData::set_center);
    ClassDB::bind_method(D_METHOD("get_control_points"), &KasaneDeformerData::get_control_points);
    ClassDB::bind_method(D_METHOD("set_control_points", "value"), &KasaneDeformerData::set_control_points);
    ClassDB::bind_method(D_METHOD("update_rotation", "center", "angle"),
                         &KasaneDeformerData::update_rotation);
    ClassDB::bind_method(D_METHOD("update_control_points", "points"),
                         &KasaneDeformerData::update_control_points);
    ClassDB::bind_method(D_METHOD("bind_to", "parent"), &KasaneDeformerData::bind_to);
    ADD_PROPERTY(PropertyInfo(Variant::STRING, "id"), "", "get_id");
    ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "angle_degrees"), "set_angle_degrees", "get_angle_degrees");
    ADD_PROPERTY(PropertyInfo(Variant::VECTOR2, "center"), "set_center", "get_center");
    ADD_PROPERTY(PropertyInfo(Variant::PACKED_VECTOR2_ARRAY, "control_points"), "set_control_points",
                 "get_control_points");
}

void KasaneDeformerData::attach(uint64_t owner_id, uint64_t generation, const String &id) {
    owner_ = owner_id;
    generation_ = generation;
    id_ = id;
}

KasaneDocumentBridge *KasaneDeformerData::owner() const {
    if (OS::get_singleton()->get_thread_caller_id() != OS::get_singleton()->get_main_thread_id())
        return nullptr;
    auto *bridge = Object::cast_to<KasaneDocumentBridge>(ObjectDB::get_instance(owner_));
    return bridge && bridge->generation() == generation_ ? bridge : nullptr;
}

bool KasaneDeformerData::is_valid() const {
    auto *b = owner();
    return b && bool(b->get_deformer_snapshot(id_)["ok"]);
}

Dictionary KasaneDeformerData::snapshot() const {
    auto *b = owner();
    return b ? b->get_deformer_snapshot(id_) : error("STALE_HANDLE", "Document was replaced or closed.");
}

double KasaneDeformerData::get_angle_degrees() const {
    return snapshot().get("angle_degrees", 0.0);
}

Vector2 KasaneDeformerData::get_center() const {
    return snapshot().get("center", Vector2());
}

PackedVector2Array KasaneDeformerData::get_control_points() const {
    return snapshot().get("control_points", PackedVector2Array());
}

Dictionary KasaneDeformerData::update_rotation(Vector2 center, double angle) {
    auto *b = owner();
    return b ? b->set_rotation(id_, center, angle)
             : error("STALE_HANDLE", "Document was replaced or closed.");
}

Dictionary KasaneDeformerData::update_control_points(const PackedVector2Array &points) {
    auto *b = owner();
    return b ? b->set_warp_points(id_, points) : error("STALE_HANDLE", "Document was replaced or closed.");
}

Dictionary KasaneDeformerData::bind_to(const String &parent) {
    auto *b = owner();
    return b ? b->set_deform_parent(id_, parent) : error("STALE_HANDLE", "Document was replaced or closed.");
}

void KasaneDeformerData::set_angle_degrees(double angle) {
    auto r = update_rotation(get_center(), angle);
    ERR_FAIL_COND_MSG(!bool(r["ok"]), String(r["message"]));
}

void KasaneDeformerData::set_center(Vector2 center) {
    auto r = update_rotation(center, get_angle_degrees());
    ERR_FAIL_COND_MSG(!bool(r["ok"]), String(r["message"]));
}

void KasaneDeformerData::set_control_points(const PackedVector2Array &points) {
    auto r = update_control_points(points);
    ERR_FAIL_COND_MSG(!bool(r["ok"]), String(r["message"]));
}
} // namespace kasane_gd

File: modules/kasane-godot/src/deformer_data.rs
use godot::obj::InstanceId;
use godot::prelude::*;

use crate::conversions::{error_dict, is_main_thread, Dictionary};
use crate::document_bridge::KasaneDocumentBridge;

#[derive(GodotClass)]
#[class(init, base=RefCounted)]
pub struct KasaneDeformerData {
    base: Base<RefCounted>,
    owner_id: u64,
    generation: u64,
    epoch: u64,
    id: GString,
}

#[godot_api]
impl KasaneDeformerData {
    pub fn attach(&mut self, owner: u64, generation: u64, epoch: u64, id: GString) {
        self.owner_id = owner;
        self.generation = generation;
        self.epoch = epoch;
        self.id = id;
    }

    fn owner(&self) -> Option<Gd<KasaneDocumentBridge>> {
        if !is_main_thread() {
            return None;
        }
        let instance_id = InstanceId::try_from_i64(self.owner_id as i64)?;
        let bridge = Gd::<KasaneDocumentBridge>::try_from_instance_id(instance_id).ok()?;
        if bridge.bind().generation() == self.generation
            && bridge.bind().object_epoch(&self.id.to_string()) == self.epoch
        {
            Some(bridge)
        } else {
            None
        }
    }

    #[func]
    pub fn is_valid(&self) -> bool {
        let Some(bridge) = self.owner() else {
            return false;
        };
        let snap = bridge.bind().get_deformer_snapshot(self.id.clone());
        snap.get("ok")
            .and_then(|v| v.try_to::<bool>().ok())
            .unwrap_or(false)
    }

    #[func]
    pub fn get_id(&self) -> GString {
        self.id.clone()
    }

    #[func]
    pub fn snapshot(&self) -> Dictionary {
        let Some(bridge) = self.owner() else {
            return error_dict("STALE_HANDLE", "Document was replaced or closed.");
        };
        let res = bridge.bind().get_deformer_snapshot(self.id.clone());
        res
    }

    #[func]
    pub fn get_angle_degrees(&self) -> f64 {
        self.snapshot()
            .get("angle_degrees")
            .and_then(|v| v.try_to::<f64>().ok())
            .unwrap_or(0.0)
    }

    #[func]
    pub fn set_angle_degrees(&mut self, angle: f64) {
        let r = self.update_rotation(self.get_center(), angle);
        let ok = r
            .get("ok")
            .and_then(|v| v.try_to::<bool>().ok())
            .unwrap_or(false);
        if !ok {
            let msg = r
                .get("message")
                .and_then(|v| v.try_to::<GString>().ok())
                .unwrap_or_default();
            godot_error!("{}", msg);
        }
    }

    #[func]
    pub fn get_center(&self) -> Vector2 {
        self.snapshot()
            .get("center")
            .and_then(|v| v.try_to::<Vector2>().ok())
            .unwrap_or(Vector2::ZERO)
    }

    #[func]
    pub fn set_center(&mut self, center: Vector2) {
        let r = self.update_rotation(center, self.get_angle_degrees());
        let ok = r
            .get("ok")
            .and_then(|v| v.try_to::<bool>().ok())
            .unwrap_or(false);
        if !ok {
            let msg = r
                .get("message")
                .and_then(|v| v.try_to::<GString>().ok())
                .unwrap_or_default();
            godot_error!("{}", msg);
        }
    }

    #[func]
    pub fn get_control_points(&self) -> PackedVector2Array {
        self.snapshot()
            .get("control_points")
            .and_then(|v| v.try_to::<PackedVector2Array>().ok())
            .unwrap_or_default()
    }

    #[func]
    pub fn set_control_points(&mut self, points: PackedVector2Array) {
        let r = self.update_control_points(points);
        let ok = r
            .get("ok")
            .and_then(|v| v.try_to::<bool>().ok())
            .unwrap_or(false);
        if !ok {
            let msg = r
                .get("message")
                .and_then(|v| v.try_to::<GString>().ok())
                .unwrap_or_default();
            godot_error!("{}", msg);
        }
    }

    #[func]
    pub fn update_rotation(&mut self, center: Vector2, angle: f64) -> Dictionary {
        let Some(mut bridge) = self.owner() else {
            return error_dict("STALE_HANDLE", "Document was replaced or closed.");
        };
        let res = bridge
            .bind_mut()
            .set_rotation(self.id.clone(), center, angle);
        res
    }

    #[func]
    pub fn update_control_points(&mut self, points: PackedVector2Array) -> Dictionary {
        let Some(mut bridge) = self.owner() else {
            return error_dict("STALE_HANDLE", "Document was replaced or closed.");
        };
        let res = bridge.bind_mut().set_warp_points(self.id.clone(), points);
        res
    }

    #[func]
    pub fn bind_to(&mut self, parent: GString) -> Dictionary {
        let Some(mut bridge) = self.owner() else {
            return error_dict("STALE_HANDLE", "Document was replaced or closed.");
        };
        let res = bridge.bind_mut().set_deform_parent(self.id.clone(), parent);
        res
    }
}
